### DISCUSSION TITLE
How to deploy to Azure Devops in a CI/CD pipeline?

### DISCUSSION THREAD

**Discussion by rapritchard** [vercel/next.js#18542] [Help]
Hi how do I go about deploying a Next JS App on an Azure Web App?

This is not the static web service that azure offer for Next JS that I am after.

Do I need to have a custom server.js file or is it possible to host a project without that on an azure web app?

Thanks

**Comment by singhparveen** [comment]
Hi @rapritchard,
I came across the same issue last week and wrote down detailed steps on my blog.
https://parveensingh.com/next-js-deployment-on-azure-app-service/
It should help you deploy the Next.JS app to Azure App Service with or without CI/CD Pipeline.

Let me know if that helps.
Cheers

**Reply by AbdullahUsmani** [reply]
You might also need to restart the AppService once.

**Comment by rapritchard** [comment]
Hi @singhparveen 

That guide hasn't worked for me. I am using a CI/CD pipeline. The app deploys to the release pipeline but I then get "The page cannot be displayed because an internal server error has occurred." on my apps URL.

I can't for the life of me figure out how to host my Next JS Project on Azure Dev Ops using the basic next start command or using a custom server + web.config.

![screenshot1](https://user-images.githubusercontent.com/32645551/99180845-bdffe400-2721-11eb-8fe5-685305856376.PNG)
![screenshot2](https://user-images.githubusercontent.com/32645551/99180846-be987a80-2721-11eb-98d6-cc66bb341cc8.PNG)

**Reply by singhparveen** [reply]
@rapritchard 
Looking at you screenshot, I don't see you doing the npm build anywhere. Internal server error is usually due to either incorrect node version defined on server or the application not built before deploying.
If you follow along step-by-step on the blog, I don't see why it wouldn't work.
I'm still running live website here using Azure DevOps: https://github.com/singhparveen/nextjs-cosmosdb

**Reply by rapritchard** [reply]
I have a post-install script that runs the build after installing.

**Comment by ArdenMendoza** [comment]
Any news on this guys? 

I am also in the same shoes as @singhparveen 

I'm getting error 500. 

@rapritchard I have read your blog and sadly it has not helped me to solve the issue. Its a very informative blog though.

**Reply by rapritchard** [reply]
Still working on getting this to work. @ArdenMendoza 

The initial problem was due to a tab in my web.config file, but I still got a server error as it seems for some reason IIS didn't understand how to run a node project.

**Reply by ArdenMendoza** [reply]
I just got mine to work. I included all the source code in the build artifacts, forwarded it to the release pipeline, then deployed it to my azure web app. (Although I don't like it, but it gets the app deployed nonetheless.) I have to figure out a way to deploy just the .next folder though.

- Scripts:
>   "scripts": {
    "dev": "next",
    "build": "next build",
    "start": "next start",
    "type-check": "tsc"
  },

- Server.js

> const { createServer } = require("http");
// const express = require('express') (Only if you app uses express)
const next = require("next");
const port = process.env.PORT || 3000;
const dev = process.env.NODE_ENV !== "production";
const app = next({ dev });
const handle = app.getRequestHandler();
app.prepare().then(() => {
  createServer((req, res) => {
    handle(req, res);
  }).listen(port, (err) => {
    if (err) throw err;
    console.log(`> Ready on http://localhost:${port}`);
  });
});

Web.config: [https://github.com/tjanczuk/iisnode/blob/master/src/samples/configuration/web.config](https://github.com/tjanczuk/iisnode/blob/master/src/samples/configuration/web.config)

**Reply by marko-hologram** [reply]
> 
> 
> I just got mine to work. I included all the source code in the build artifacts, forwarded it to the release pipeline, then deployed it to my azure web app. (Although I don't like it, but it gets the app deployed nonetheless.) I have to figure out a way to deploy just the .next folder though.
> 
>     * Scripts:
> 
> 
> > "scripts": {
> > "dev": "next",
> > "build": "next build",
> > "start": "next start",
> > "type-check": "tsc"
> > },
> 
>     * Server.js
> 
> 
> > const { createServer } = require("http");
> > // const express = require('express') (Only if you app uses express)
> > const next = require("next");
> > const port = process.env.PORT || 3000;
> > const dev = process.env.NODE_ENV !== "production";
> > const app = next({ dev });
> > const handle = app.getRequestHandler();
> > app.prepare().then(() => {
> > createServer((req, res) => {
> > handle(req, res);
> > }).listen(port, (err) => {
> > if (err) throw err;
> > console.log(`> Ready on http://localhost:${port}`);
> > });
> > });
> 
> Web.config: https://github.com/tjanczuk/iisnode/blob/master/src/samples/configuration/web.config

I'm quite sure you cannot just deploy `.next` folder because "start" script uses `next start` command and `next` then has to come from `node_modules`. There is maybe a way to slim down `node_modules`, but I'm not sure how.

**Reply by rapritchard** [reply]
I've managed to finally deploy it, the web config you provided @ArdenMendoza seemed to provide the missing link.
Here is my server.js, web.config for those who struggle.

Web.config
```
<?xml version="1.0" encoding="utf-8"?>
<!--	
     This configuration file is required if iisnode is used to run node processes behind	
     IIS or IIS Express.  For more information, visit:	
     https://github.com/tjanczuk/iisnode/blob/master/src/samples/configuration/web.config	
-->	
<configuration>
  <system.webServer>
    <!-- Visit http://blogs.msdn.com/b/windowsazure/archive/2013/11/14/introduction-to-websockets-on-windows-azure-web-sites.aspx for more information on WebSocket support -->	
    <webSocket enabled="false" />
    <handlers>
      <!-- Indicates that the server.js file is a node.js site to be handled by the iisnode module -->
      <add name="iisnode" path="server.js" verb="*" modules="iisnode"/>
    </handlers>
    <rewrite>
      <rules>
        <!-- Do not interfere with requests for node-inspector debugging -->
        <rule name="NodeInspector" patternSyntax="ECMAScript" stopProcessing="true">
          <match url="^server.js\/debug[\/]?" />
        </rule>
        <!-- First we consider whether the incoming URL matches a physical file in the /public folder -->
        <rule name="StaticContent">
          <action type="Rewrite" url="public{REQUEST_URI}"/>
        </rule>
        <!-- All other URLs are mapped to the node.js site entry point -->
        <rule name="DynamicContent">
          <conditions>
            <add input="{REQUEST_FILENAME}" matchType="IsFile" negate="True"/>
          </conditions>
          <action type="Rewrite" url="server.js"/>
        </rule>
      </rules>
    </rewrite>
    <!-- 'bin' directory has no special meaning in node.js and apps can be placed in it -->	
    <security>
      <requestFiltering>
        <hiddenSegments>
          <remove segment="bin"/>
        </hiddenSegments>
      </requestFiltering>
    </security>
    <!-- Make sure error responses are left untouched -->
    <httpErrors existingResponse="PassThrough" />
    <!--
      You can control how Node is hosted within IIS using the following options:	
        * watchedFiles: semi-colon separated list of files that will be watched for changes to restart the server	
        * node_env: will be propagated to node as NODE_ENV environment variable	
        * debuggingEnabled - controls whether the built-in debugger is enabled	
      See https://github.com/tjanczuk/iisnode/blob/master/src/samples/configuration/web.config for a full list of options	
    -->
    <iisnode
      node_env="%node_env%"
      nodeProcessCountPerApplication="1"
      maxConcurrentRequestsPerProcess="1024"
      maxNamedPipeConnectionRetry="100"
      namedPipeConnectionRetryDelay="250"
      maxNamedPipeConnectionPoolSize="512"
      maxNamedPipePooledConnectionAge="30000"
      asyncCompletionThreadCount="0"
      initialRequestBufferSize="4096"
      maxRequestBufferSize="65536"
      watchedFiles="web.config;*.js;iisnode.yml"
      uncFileChangesPollingInterval="5000"
      gracefulShutdownTimeout="60000"
      loggingEnabled="true"
      logDirectory="iisnode"
      debuggingEnabled="true"
      debugHeaderEnabled="false"
      debuggerPortRange="5058-6058"
      debuggerPathSegment="debug"
      maxLogFileSizeInKB="128"
      maxTotalLogFileSizeInKB="1024"
      maxLogFiles="20"
      devErrorsEnabled="true"
      flushResponse="false"
      enableXFF="false"
      promoteServerVars=""
      configOverrides="iisnode.yml"
     />
  </system.webServer>
</configuration>
```
Server.js
```javascript
const express = require("express");
const next = require("next");

const dev = process.env.NODE_ENV !== "production";
const port = process.env.PORT || 3000;
const app = next({ dev });
const handle = app.getRequestHandler();

app.prepare().then(() => {
  const server = express();

  server.all("*", (req, res) => {
    return handle(req, res);
  });

  server.listen(port, (err) => {
    if (err) throw err;
    console.log(`> Ready on http://localhost:${port}`);
  });
});
```


All I need to do now is get the environment variable to be set to production in azure and i'm happy :)

**Comment by jonreast** [comment]
Great article! Two issues I ran into when setting up the ADO release tasks. 

**Set node version to a value Azure AppService didn't have available.** 
Solution: 
1. Verified the version of node installed on the Azure WebApp using Kudo via Advanced Tools in the Azure Portal
2. Set the version of the AppService via Application Settings key WEBSITE_NODE_DEFAULT_VERSION to ~14 (this allowed me to use the latest v14 without having to specify the exact verstion)
3. Set the platform = x64 (per the node version installed discovered in step 1) 

**Remove node version and any start commands from ADO deployment tasks**
Solution: 
1. Started with ADO release template "Deploy a Node.js app to Azure App Services"
2. Connected my Azure subscription/appservice 
3. Remove application setting node version value ADO generated (as it's already defined in Azure)

**Tip**
If your unfamiliar with web.config/server.js please note these will inform Azure WebApps how to start the application, no post-deployment commands are necessary.

**Comment by RodolphoFernandes** [comment]
Good afternoon!

I managed to solve my problem.

add the two files server.js and web.config and in package.json change the scripts

"dev": "node server.js",
"build": "next build",
"start": "node server.js"
